### PR TITLE
html graph connectivity

### PR DESCRIPTION
**Pull request by haim-kermany**
html connectivity.
Ziv, please review the drawio package
Shiri & Adi - please check the rest

PLEASE notice:
1. we will have another review on the html UI (visibility, usability, correctness...)
2. For now, the term "drawio" is used for all graphical solutions (for example  drawioGenerator.go). renaing need to be done.
I didnt do massive renaming, to easy the CR

### CODE REVIEW

File: pkg/ibmvpc/ibmDrawioGenerator.go [hunk 1]
```diff
@@ -10,6 +10,7 @@ func (v *VPC) ShowOnSubnetMode() bool                  { return true }
 func (z *Zone) ShowOnSubnetMode() bool                 { return true }
 func (s *Subnet) ShowOnSubnetMode() bool               { return true }
 func (sgl *SecurityGroupLayer) ShowOnSubnetMode() bool { return false }
+func (sg *SecurityGroup) ShowOnSubnetMode() bool       { return false }
```
Comment (ShiriMoran) [inline pkg/ibmvpc/ibmDrawioGenerator.go:13]: Why was this added?

**Review by ShiriMoran — COMMENTED** [review]
LGTM

File: pkg/vpcmodel/drawioOutput.go [hunk 2]
```diff
@@ -32,8 +32,16 @@ type DrawioOutputFormatter struct {
 	nodeRouters     map[drawio.TreeNodeInterface]drawio.IconTreeNodeInterface
 	multiVpcRouters map[string]drawio.IconTreeNodeInterface
 	uc              OutputUseCase
+	outFormat       OutFormat
 }
 
+func newDrawioOutputFormatter(outFormat OutFormat) *DrawioOutputFormatter {
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/vpcmodel/drawioOutput.go:38]: 1. now we have a parameter passed in creation
2. it make more sense to have a contractor than to initialize members in the init function

File: pkg/vpcmodel/output.go [hunk 3]
```diff
@@ -66,7 +71,9 @@ func NewOutputGenerator(c1, c2 MultipleVPCConfigs, grouping bool, uc OutputUseCa
 		nodesConn:      map[string]*VPCConnectivity{},
 		subnetsConn:    map[string]*VPCsubnetConnectivity{},
 	}
-	if !archOnly {
+	graphicFormat := slices.Contains([]OutFormat{DRAWIO, ARCHDRAWIO, SVG, ARCHSVG, HTML, ARCHHTML}, f)
+	archOnlyFormat := slices.Contains([]OutFormat{ARCHDRAWIO, ARCHSVG, ARCHHTML}, f)
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/vpcmodel/output.go:75]: if we will have a non graphical architectural format that, it will be part of the list

File: pkg/ibmvpc/ibmDrawioGenerator.go [hunk 4]
```diff
@@ -44,11 +49,15 @@ func (s *Subnet) GenerateDrawioTreeNode(gen *vpcmodel.DrawioGenerator) drawio.Tr
...
-	tn := drawio.NewSGTreeNode(gen.TreeNode(sgl.VPC()).(*drawio.VpcTreeNode), sgl.Name())
 	for _, sg := range sgl.sgList {
-		for _, member := range sg.members {
-			tn.AddIcon(gen.TreeNode(member).(drawio.IconTreeNodeInterface))
-		}
+		gen.TreeNode(sg)
+	}
+	return nil
+}
+func (sg *SecurityGroup) GenerateDrawioTreeNode(gen *vpcmodel.DrawioGenerator) drawio.TreeNodeInterface {
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/ibmvpc/ibmDrawioGenerator.go:57]: ibmDrawioGenerator.go mostly contains  GenerateDrawioTreeNode() functions for all resources defined in ibmvpc package.
all  GenerateDrawioTreeNode()  creates a drawio treeNode for the resource.
it is well documented at drawioGenerator.go
it has nothing to do with interactivity, I just fix a bug GenerateDrawioTreeNode() of SG

File: pkg/drawio/treeNodeInterface.go [hunk 5]
```diff
@@ -126,6 +149,20 @@ func absoluteGeometry(tn TreeNodeInterface) (x, y int) {
 		tn.Y() + tn.DrawioParent().Location().firstRow.y() + tn.DrawioParent().Location().yOffset
 }
 
+func joinLabels(labels []string, sep string) string {
+	labels2 := []string{}
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/drawio/treeNodeInterface.go:153]: this method if joining labels.
the name describe what it does 
(like sort(), join() ... )

File: pkg/drawio/abstractTreeNode.go [hunk 6]
```diff
@@ -24,18 +22,13 @@ type abstractTreeNode struct {
...
-	labels2 := []string{}
-	for _, l := range labels {
-		if l != "" {
-			labels2 = append(labels2, l)
-		}
-	}
-	return strings.Join(labels2, "&#xa;")
-}
+func (tn *abstractTreeNode) labels() []string    { return []string{tn.name} }
+func (tn *abstractTreeNode) Kind() string        { return tn.kind }
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/drawio/abstractTreeNode.go:29]: it is a string, because I need it as a string to display at the html file.
it just return the value of method the method Kind() of VPCResourceIntf

File: pkg/vpcmodel/drawioOutput.go [hunk 7]
```diff
@@ -177,10 +183,81 @@ func (d *DrawioOutputFormatter) createEdges() {
 	}
 }
 
+// explainableEndpoints() createExplanations() to be reimplement with real Shiri work
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/vpcmodel/drawioOutput.go:186]: אמן

File: pkg/ibmvpc/ibmDrawioGenerator.go (hunk 1 above)
Comment (haim-kermany) [reply pkg/ibmvpc/ibmDrawioGenerator.go:13]: every resource has a method  ShowOnSubnetMode()
do we show it on the canvas when we are of subnet mode?

File: README.md [hunk 8]
```diff
@@ -29,7 +29,7 @@ Usage of vpc-network-config-analyzer:
         File path to store resources collected from the cloud provider
   -format string
         Output format; must be one of:
-        txt, md, json, drawio, arch_drawio, debug (default "txt")
+        txt, md, json, drawio, arch_drawio, svg, arch_svg, html, arch_html, debug (default "txt")
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply README.md:32]: we might need to elaborate about the formats, but not as part of the usage
@zivnevo? @adisos ?

File: pkg/drawio/treeNodeInterface.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/drawio/treeNodeInterface.go:153]: I meant to rename `labels2`  to `labelsJoined`; the function name is indeed fine

File: pkg/ibmvpc/ibmDrawioGenerator.go (hunk 1 above)
Comment (ShiriMoran) [reply pkg/ibmvpc/ibmDrawioGenerator.go:13]: Yes, that I remembered, but why was this specific method added for `(sg *SecurityGroup)` in this specific PR?

File: pkg/vpcmodel/output.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/vpcmodel/output.go:75]: But this format is graphic, isn't it? Am I missing something?

[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/vpcmodel/output.go:75]: this code is general for all output formats.
in case of not arch - we do connectivity analysis.
for now, all are arch formats are graphic, but it isnt necessary always be the case.

File: pkg/ibmvpc/ibmDrawioGenerator.go (hunk 1 above)
Comment (haim-kermany) [reply pkg/ibmvpc/ibmDrawioGenerator.go:13]: I had a bug, in which I created a Treenode per SecurityGroupLayer, instead creating a treeNode per SecurityGroup.
with the bug, SecurityGroup did not implemented the interface DrawioResourceIntf
now it does

File: pkg/ibmvpc/ibmDrawioGenerator.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/ibmvpc/ibmDrawioGenerator.go:58]: done

File: pkg/drawio/treeNodeInterface.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/drawio/treeNodeInterface.go]: done

File: pkg/vpcmodel/drawioOutput.go [hunk 9]
```diff
@@ -177,10 +183,81 @@ func (d *DrawioOutputFormatter) createEdges() {
...
+		if !vpcConfig1.IsMultipleVPCsConfig {
+			if !subnetMode {
+				for _, n := range vpcConfig1.Nodes {
+					if !n.IsExternal() && d.showResource(n) {
+						allEndpoints = append(allEndpoints, n)
+					}
+				}
+			} else {
+				for _, s := range vpcConfig1.Subnets {
+					if !s.IsExternal() && d.showResource(s) {
```
Comment (zivnevo) [inline pkg/vpcmodel/drawioOutput.go:235]: I don't believe a Subnet can be external

File: pkg/vpcmodel/drawioOutput.go [hunk 10]
```diff
@@ -177,10 +183,81 @@ func (d *DrawioOutputFormatter) createEdges() {
...
+				dsts = *dstList
+			}
+			for _, src := range srcs {
+				for _, dst := range dsts {
+					explanations[expKey{src, dst}] = line.String()
+				}
+			}
+		}
+	}
+	explanationsList := []drawio.ExplanationEntry{}
```
Comment (zivnevo) [inline pkg/vpcmodel/drawioOutput.go:215]: preallocate memory

File: pkg/vpcmodel/drawioOutput.go [hunk 11]
```diff
@@ -177,10 +183,81 @@ func (d *DrawioOutputFormatter) createEdges() {
...
+// explainableEndpoints() createExplanations() to be reimplement with real Shiri work
+func (d *DrawioOutputFormatter) createExplanations() []drawio.ExplanationEntry {
+	type expKey struct {
+		src, dst EndpointElem
+	}
+	explanations := map[expKey]string{}
+	allEndpoints := d.explainableEndpoints()
+	for _, src := range allEndpoints {
+		for _, dst := range allEndpoints {
+			explanations[expKey{src, dst}] = "No Connectivity from " + src.Name() + " to " + dst.Name()
```
Comment (zivnevo) [inline pkg/vpcmodel/drawioOutput.go:195]: We do produce explanations for the connectivity between an internal endpoint and an external entity. Does the code take care of that?

File: pkg/drawio/treeNodeInterface.go [hunk 12]
```diff
@@ -126,6 +149,20 @@ func absoluteGeometry(tn TreeNodeInterface) (x, y int) {
 		tn.Y() + tn.DrawioParent().Location().firstRow.y() + tn.DrawioParent().Location().yOffset
 }
 
+func joinLabels(labels []string, sep string) string {
+	labelsToJoin := []string{}
+	for _, l := range labels {
+		if l != "" {
+			labelsToJoin = append(labelsToJoin, l)
+		}
+	}
```
Comment (zivnevo) [inline pkg/drawio/treeNodeInterface.go:158]:
```suggestion
	labelsToJoin := slices.Clone(labels)
	labelsToJoin = slices.DeleteFunc(labelsToJoin, func(s string) bool {return s == ""})
```

File: pkg/drawio/createMapFile.go [hunk 13]
```diff
@@ -58,52 +119,70 @@ func (data *templateData) AY(tn TreeNodeInterface) int {
...
+		&SubnetTreeNode{},
+		&SGTreeNode{},
+		&PartialSGTreeNode{},
+		&GroupSquareTreeNode{},
+	}
+	var lines, icons, orderedNodes []TreeNodeInterface
+	squaresBuckets := map[reflect.Type][]TreeNodeInterface{}
+	for _, t := range squareOrders {
+		squaresBuckets[reflect.TypeOf(t).Elem()] = []TreeNodeInterface{}
+	}
```
Comment (zivnevo) [inline pkg/drawio/createMapFile.go:148]: Is this initialization needed?
There is no problem in calling `append` with `nil` as the first argument

File: pkg/drawio/nodesRelations.go [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+		psg = append(psg, psgTn)
+	}
+	for _, icon := range sgTn.elements {
+		icons = append(icons, icon)
+	}
+	return psg, icons
+}
+
+// lineInfo is a struct with all the treeNodes representing a grouping line:
+// for example, for a connection  A,B -> C,D wi will can have the following on the canvas:
```
Comment (zivnevo) [inline pkg/drawio/nodesRelations.go:86]:
```suggestion
// for example, for a connection  A,B -> C,D we can have the following on the canvas:
```

File: pkg/drawio/nodesRelations.go [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+		info.src = src.Parent()
+		info.srcGroupingLines = src.(*GroupPointTreeNode).groupedIconsConns
+		info.srcGroupingPoint = src
+	}
+	if dstIsGP {
+		info.dst = dst.Parent()
+		info.dstGroupingLines = dst.(*GroupPointTreeNode).groupedIconsConns
+		info.dstGroupingPoint = dst
+	}
+	// if both src and dst are not grouping point, its just a regular line
```
Comment (zivnevo) [inline pkg/drawio/nodesRelations.go:122]: This comment confused me a little bit. I'm not sure it is needed.

File: pkg/drawio/nodesRelations.go [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+	res := map[string]map[string][]string{}
+	for _, node := range data.Nodes {
+		nodeID := strconv.Itoa(int(node.ID()))
+		res[nodeID] = map[string][]string{}
+		nodeRelations := []string{strconv.Itoa(int(data.RootID()))}
+		for _, n := range rel[node] {
+			nodeRelations = append(nodeRelations, strconv.Itoa(int(n.ID())))
+		}
+		res[nodeID]["relations"] = nodeRelations
+		res[nodeID]["graphExplanation"] = []string{"Connectivity graph of " + data.NodeName(node)}
```
Comment (zivnevo) [inline pkg/drawio/nodesRelations.go:28]: The string `"Connectivity graph of "` is duplicated for each and every node. Can't we have it in the JavaScript?

**Review by zivnevo — CHANGES_REQUESTED** [review]

File: cmd/analyzer/parse_args.go [hunk 17]
```diff
@@ -387,7 +390,8 @@ func errorInArgs(args *InArgs, flagset *flag.FlagSet) error {
 	if !fileForDiffSpecified && diffAnalysis {
 		return fmt.Errorf("missing parameter vpc-config-second for diff analysis %s", *args.AnalysisType)
 	}
-	if slices.Contains([]string{DRAWIOFormat, ARCHDRAWIOFormat}, *args.OutputFormat) && *args.OutputFile == "" {
+	graphicFormats := []string{DRAWIOFormat, ARCHDRAWIOFormat, SVGFormat, ARCHSVGFormat, HTMLFormat, ARCHHTMLFormat}
+	if slices.Contains(graphicFormats, *args.OutputFormat) && *args.OutputFile == "" {
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply cmd/analyzer/parse_args.go:394]: 1. the golang template request an io.Writer parameter, it does not return the output
2. we print to stdout some warnings, mix this output with the html code make no sence

File: pkg/vpcmodel/drawioOutput.go (hunk 9 above)
Comment (haim-kermany) [reply pkg/vpcmodel/drawioOutput.go]: fixed

File: pkg/vpcmodel/drawioOutput.go (hunk 10 above)
Comment (haim-kermany) [reply pkg/vpcmodel/drawioOutput.go]: done

File: pkg/vpcmodel/drawioOutput.go (hunk 11 above)
Comment (haim-kermany) [reply pkg/vpcmodel/drawioOutput.go:195]: this code do take care of that.
however, this code is about to be deleted with Shiri implementation

File: pkg/drawio/nodesRelations.go (hunk 14 above)
Comment (haim-kermany) [reply pkg/drawio/nodesRelations.go]: fixed

File: pkg/drawio/nodesRelations.go (hunk 15 above)
Comment (haim-kermany) [reply pkg/drawio/nodesRelations.go]: comment elaborated

File: pkg/drawio/nodesRelations.go (hunk 16 above)
Comment (haim-kermany) [reply pkg/drawio/nodesRelations.go:28]: yes.
however, I assume that in the visibility review to the html files, the graphExplanation will be more complicated.
I want to implement this complication here, and not in the javascript

File: cmd/analyzer/parse_args.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (zivnevo) [reply cmd/analyzer/parse_args.go:394]: 1. You can give it a `bytes.Buffer` than use [Buffer.String()](https://pkg.go.dev/bytes@go1.22.2#Buffer.String)
2. Warnings should go to stderr, output should go to stdout.  Maybe this only works after Issue #376 is closed.

[the remark this answers is not in the document]
Comment (zivnevo) [reply cmd/analyzer/parse_args.go:394]: Can wait for a later PR

**Review by zivnevo — APPROVED** [review]
